### ISSUE TITLE
Support maps and sets

### ISSUE THREAD

**Issue by gbrail** [mozilla/rhino#159]
ES6 adds:

Map
Set
WeakMap
WeakSet

these are supported by recent versions of ES6.

**Comment by eshepelyuk** [comment]
Hello @gbrail 
I am working on this issue. Could you please add corresponding labels ?
During implementation I had to implement naive support for `Symbol` as well to conform to `test262` suite.
I would like to create separate PR with simple `Symbol` implementation to unblock other `ES6` tasks like #164 and maybe #162 before implementing this task.
What do you think ?

**Comment by gbrail** [comment]
Absolutely -- but as you said, a lot of stuff depends on Symbol. So it'd make sense to get Symbol built first, in a separate PR.

**Comment by eshepelyuk** [comment]
I plan to submit PR after 1.7.7 release. Is it OK ?

**Comment by gbrail** [comment]
Yep -- thanks!

On Wed, Jun 17, 2015 at 9:51 AM, Evgeny Shepelyuk <notifications@github.com>
wrote:

> I plan to submit PR after 1.7.7 release. Is it OK ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mozilla/rhino/issues/159#issuecomment-112874841>.
>



-- 
*Greg Brail* | *apigee <https://apigee.com/>* | twitter @gbrail
<http://twitter.com/gbrail> @apigee <https://twitter.com/apigee>
<http://iloveapis.com/>

**Comment by eshepelyuk** [comment]
Hello @gbrail 
What do you think about integrating [Apache Commons Collections 4.0](https://commons.apache.org/proper/commons-collections/) into `Rhino` project ?
I would like to use it classes for implementing ES6 iteration, where, according to spec, changes applied to map or set should be visible to iterator even after iterator creation. 
Such behaviour is not supported in standard JDK classes and implementing it looks really ugly and verbose.

**Comment by gbrail** [comment]
I think that we should stay away from including a dependency. Rhino is used
in a lot of different contexts and if we end up with a dependency on this
library, then it is likely (guaranteed, actually) that we will end up with
a conflict with some other project that depends on a different version.
Rhino is a pretty low-level component.

If the requirements are different, I don't see a problem with Rhino having
its own Map and Set implementations designed expressly for the needs of
JavaScript.

FWIW, a lot of the semantics of a map are already implemented in
ScriptableObject, via the slot array.

On Mon, Jun 29, 2015 at 1:12 PM, Evgeny Shepelyuk <notifications@github.com>
wrote:

> Hello @gbrail <https://github.com/gbrail>
> What do you think about integrating Apache Commons Collections 4.0
> <https://commons.apache.org/proper/commons-collections/> into Rhino
> project ?
> I would like to use it classes for implementing ES6 iteration, where,
> according to spec, changes applied to map or set should be visible to
> iterator even after iterator creation.
> Such behaviour is not supported in standard JDK classes and implementing
> it looks really ugly and verbose.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mozilla/rhino/issues/159#issuecomment-116826184>.
>



-- 
*Greg Brail* | *apigee <https://apigee.com/>* | twitter @gbrail
<http://twitter.com/gbrail> @apigee <https://twitter.com/apigee>
<http://iloveapis.com/>

**Comment by gbrail** [comment]
Implemented in https://github.com/mozilla/rhino/pull/472.

WeakMap and WeakSet are coming soon.

**Comment by rbri** [comment]
this is done also
